lib/engine-jit/src/code_memory.rs

```diff
@@ -17,193 +13,144 @@ use wasmer_vm::{FunctionBodyPtr, Mmap, VMFunctionBody};
-
-            result.push(FunctionBodyPtr(vmfunc as *mut [VMFunctionBody]));
-
-            padding = get_align_padding_size(next_start as usize, ARCH_FUNCTION_ALIGNMENT);
-            start = next_start;
-            buf = next_buf;
-        }
+        // 2. Allocate the pages. Mark them all read-write.
 
-        Ok(result)
-    }
+        self.mmap = Mmap::with_at_least(total_len)?;
 
-    /// Allocate a continuous memory block for a single compiled function.
-    /// TODO: Reorganize the code that calls this to emit code directly into the
-    /// mmap region rather than into a Vec that we need to copy in.
-    pub fn allocate_for_function(
-        &mut self,
-        registry: &mut UnwindRegistry,
-        func: &FunctionBody,
-    ) -> Result<&mut [VMFunctionBody], String> {
-        let size = Self::function_allocation_size(func);
-
-        let (buf, start) = self.allocate(size, ARCH_FUNCTION_ALIGNMENT)?;
-        let base_address = buf.as_ptr() as usize - start;
-
-        let (_, _, vmfunc) = Self::copy_function(registry, base_address, func, start as u32, buf);
-        assert!(vmfunc as *mut _ as *mut u8 as usize % ARCH_FUNCTION_ALIGNMENT == 0);
-
-        Ok(vmfunc)
-    }
-
-    /// Allocate a continuous memory block for an executable custom section.
-    pub fn allocate_for_executable_custom_section(
-        &mut self,
-        section: &SectionBody,
-    ) -> Result<&mut [u8], String> {
-        let section = section.as_slice();
-        let (buf, _) = self.allocate(section.len(), ARCH_FUNCTION_ALIGNMENT)?;
-        buf.copy_from_slice(section);
-        Ok(buf)
-    }
+        // 3. Determine where the pointers to each function, executable section
+        // or data section are. Copy the functions. Collect the addresses of each and return them.
 
-    /// Allocate a continuous memory block for a readable custom section.
-    pub fn allocate_for_custom_section(
-        &mut self,
-        section: &SectionBody,
-    ) -> Result<&mut [u8], String> {
-        let section = section.as_slice().to_vec();
-        self.read_sections.push(section);
-        Ok(self
-            .read_sections
-            .last_mut()
-            .ok_or_else(|| "Can't get last section".to_string())?)
-    }
+        let mut bytes = 0;
+        let mut buf = self.mmap.as_mut_slice();
+        for func in functions {
+            let len = round_up(
+                Self::function_allocation_size(func),
+                ARCH_FUNCTION_ALIGNMENT,
+            );
+            let (func_buf, next_buf) = buf.split_at_mut(len);
+            buf = next_buf;
+            bytes += len;
 
-    /// Make all allocated memory executable.
-    pub fn publish(&mut self) {
-        self.push_current(0)
-            .expect("failed to push current memory map");
+            let vmfunc = Self::copy_function(&mut self.unwind_registry, func, func_buf);
+            assert!(vmfunc as *mut _ as *mut u8 as usize % ARCH_FUNCTION_ALIGNMENT == 0);
```

Comment:
You can probably just cast this directly to a usize once it's a pointer, I don't think it matters that it be a `*mut u8` first

Reply:
Also consider using `assert_eq!` which will print out what the value is if it's not equal, whereas normal assert will only say that the given boolean expression failed